src/QirTools/pyqir/qirlib/src/emit/mod.rs

```diff
@@ -0,0 +1,276 @@
+use self::constants::Constants;
+use self::intrinsics::Intrinsics;
+use self::runtime_library::RuntimeLibrary;
+use self::types::Types;
+use inkwell::values::BasicValueEnum;
+
+use super::interop::SemanticModel;
+use std::collections::BTreeMap;
+use std::path::Path;
+
+pub mod constants;
+    mod intrinsics;
+pub mod qir;
+mod runtime_library;
+pub mod types;
+
+pub struct Emitter {}
+impl Emitter {
+    pub fn write(model: &SemanticModel, file_name: &str) -> Result<(), String> {
+        let ctx = inkwell::context::Context::create();
+        let context = Context::new(&ctx, model.name.as_str());
+
+        Emitter::build_entry_function(&context, model)?;
+        context.emit_ir(file_name);
+
+        Ok(())
+    }
+
+    pub fn get_ir_string(model: &SemanticModel) -> Result<String, String> {
+        let ctx = inkwell::context::Context::create();
+        let context = Context::new(&ctx, model.name.as_str());
+
+        Emitter::build_entry_function(&context, model)?;
+
+        Ok(context.get_ir_string())
+    }
+
+    pub fn build_entry_function(context: &Context<'_>, model: &SemanticModel) -> Result<(), String> {
+        let entrypoint = qir::get_entry_function(context);
+
+        let entry = context.context.append_basic_block(entrypoint, "entry");
+        context.builder.position_at_end(entry);
+
+        let qubits = Emitter::write_qubits(&model, context);
+
+        let registers = Emitter::write_registers(&model, context);
+
+        Emitter::write_instructions(&model, context, &qubits, &registers);
+
+        Emitter::free_qubits(context, &qubits);
+
+        let output = registers.get("results").unwrap();
+        context.builder.build_return(Some(&output.0));
+
+        if let Err(err) = context.module.verify() {
+            let message = err.to_string();
+            return Err(message);
+        }
+        Ok(())
+    }
+
+    fn free_qubits<'ctx>(context: &Context<'ctx>, qubits: &BTreeMap<String, BasicValueEnum<'ctx>>) {
```

Comment:
Purely out of curiosity, what motivates the use of a BTreeMap versus a HashMap here?